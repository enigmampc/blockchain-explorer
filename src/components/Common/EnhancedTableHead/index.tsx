import React from 'react'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import TableCell from '@material-ui/core/TableCell'
import TableSortLabel from '@material-ui/core/TableSortLabel'
import styled from 'styled-components'

export enum FlexAlign {
  'flexStart' = 'flex-start',
  'flexEnd' = 'flex-end',
  'center' = 'center',
}

export type HeaderCell = {
  id: string
  useClassShowOnDesktop: boolean
  align: FlexAlign
  label: string
}

export interface EnhancedTableHeadProps extends React.HTMLAttributes<HTMLDivElement> {
  headerCells: HeaderCell[]
  order: 'asc' | 'desc'
  orderBy: string
  onRequestSort: CallableFunction
}

const TableHeadText = styled.span`
  color: #999;
  font-size: 12px;
  white-space: nowrap;

  & + svg {
    opacity: 1;
  }
`

const EnhancedTableHead = ({ headerCells, order, orderBy, onRequestSort }: EnhancedTableHeadProps) => {
  const createSortHandler = (property: string) => (event: React.SyntheticEvent) => {
    onRequestSort(event, property)
  }

  return (
    <TableHead>
      <TableRow>
        {headerCells.map(headerCell => {
          const active = orderBy === headerCell.id

          return (
            <TableCell
              key={headerCell.id}
              sortDirection={active ? order : false}
              className={headerCell.useClassShowOnDesktop ? 'showOnDesktop' : ''}
              style={{
                paddingRight: '5px',
                display: headerCell.useClassShowOnDesktop ? 'none' : 'table-cell',
              }}
            >
              {active ? (
                <TableSortLabel
                  active={active}
                  direction={order}
                  hideSortIcon={false}
                  onClick={createSortHandler(headerCell.id)}
                  style={{
                    justifyContent: headerCell.align,
                    display: 'flex',
                  }}
                >
                  <TableHeadText>{headerCell.label}</TableHeadText>
                </TableSortLabel>
              ) : (
                <TableHeadText
                  onClick={createSortHandler(headerCell.id)}
                  style={{
                    justifyContent: headerCell.align,
                    display: 'flex',
                    cursor: 'pointer',
                  }}
                >
                  {headerCell.label}
                </TableHeadText>
              )}
            </TableCell>
          )
        })}
      </TableRow>
    </TableHead>
  )
}

export default EnhancedTableHead
