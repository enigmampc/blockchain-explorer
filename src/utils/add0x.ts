const add0x = (hex: string): string => (hex ? (hex.indexOf('0x') === 0 ? hex : `0x${hex}`) : '')

export default add0x
