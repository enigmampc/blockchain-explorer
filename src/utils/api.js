import { formatNumber } from './numeralFormats'

export const cryptoCompare = async () => {
  const url = new URL('https://min-api.cryptocompare.com/data/pricemultifull')
  const params = {
    fsyms: 'ENG',
    tsyms: 'USD',
  }
  url.search = new URLSearchParams(params).toString()
  const {
    RAW: {
      ENG: {
        USD: { PRICE, MKTCAP },
      },
    },
  } = await (await fetch(url)).json().catch(console.error)

  return {
    price: formatNumber(PRICE, 4),
    marketCap: formatNumber(MKTCAP, 0),
  }
}

export const coinMarketCap = async () => {
  const url = new URL('https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest')
  const params = {
    id: 2044,
  }
  url.search = new URLSearchParams(params).toString()
  const headers = new Headers({
    'X-CMC_PRO_API_KEY': process.env.REACT_APP_COINMARKET_CAP_API_KEY,
  })
  const init = {
    method: 'GET',
    headers,
    cache: 'default',
  }
  const request = new Request(url, init)
  const response = await (await fetch(request)).json()

  if (response.error_code) throw new Error(response.error_message)

  const {
    data: {
      [params.id]: {
        quote: {
          USD: { market_cap, price },
        },
      },
    },
  } = response

  return {
    price: formatNumber(price, 4),
    marketCap: formatNumber(market_cap, 0),
  }
}

export const binance = async () => {
  const binanceURL = 'https://api.binance.com/api/v1/ticker/price?symbol='
  const { price } = await (await fetch(`${binanceURL}ENGBTC`)).json().catch()

  return {
    price: formatNumber(price, 4),
    marketCap: '',
  }
}
