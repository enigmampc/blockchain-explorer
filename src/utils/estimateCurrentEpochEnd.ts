import ethApi from './eth'
import { EpochProps } from '../components/Epochs'

const calculateRangesDurations = (range: number[]) => {
  const acc = []

  for (let i = 0; i < range.length; i += 2) {
    const rangeElement = range[i]
    acc.push(rangeElement - range[i + 1])
  }

  return acc
}

const distanceMedian = (range: number[]) => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (length === 0) return length

  durations.sort((a, b) => a - b)

  const half = Math.floor(length / 2)

  return length % 2 ? durations[half] : Math.floor((durations[half - 1] + durations[half]) / 2)
}

const distanceAverage = (range: number[]) => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (!length) return 0

  const sum = durations.reduce((a, b) => a + b)
  return Math.floor(sum / length)
}

export interface EstimatesCurrentEpochEnd {
  pendingTime: number
  finishBlockNumber: number
}

const estimateCurrentEpochEnd = async (epochs?: EpochProps[], average = false): Promise<EstimatesCurrentEpochEnd> => {
  let pendingTime = 0 // estimated life time for the current epoch
  let finishBlockNumber = 0 // estimated block for when the current epoch may end

  if (!epochs) return { pendingTime, finishBlockNumber }

  const sortedEpochs = epochs.sort((a, b) => +b.id - +a.id)
  const currentEpoch = sortedEpochs.slice(0, 1)[0] // currently active epoch

  // builds an array of the form [10, 1, 21, 11, 40, 22]
  // where 10 is the last block of the first epoch in the collection
  // and 1 is the first block of the first epoch in the collection
  // 21 is the last block of the second epoch and 11 is it's first block and so on
  const blocksRange = sortedEpochs
    .reduce((acc: number[], epoch, index) => {
      if (index !== 0) {
        acc.push(+epoch.startBlockNumber)
        acc.push(+epoch.endBlockNumber)
      }

      return acc
    }, [])
    .reverse()

  try {
    const [currentBlock, epochsRangesTimestamps] = await Promise.all([
      ethApi.getBlockNumber(),
      ethApi.getBatchBlocksTimestamps(blocksRange),
    ])
    const epochBlockCount = average ? distanceAverage(blocksRange) : distanceMedian(blocksRange)
    const epochDuration = average ? distanceAverage(epochsRangesTimestamps) : distanceMedian(epochsRangesTimestamps)
    finishBlockNumber = +currentEpoch.startBlockNumber + epochBlockCount
    const currentBlocksLeft = finishBlockNumber - +currentBlock
    const currentBlocksTimes = epochDuration / epochBlockCount
    pendingTime = Math.floor(currentBlocksLeft * currentBlocksTimes) * 1000 // milliseconds
  } catch (e) {
    console.error('Failed to retrieve blocks tiemstamps', e)
  }

  return { pendingTime, finishBlockNumber }
}

export default estimateCurrentEpochEnd
