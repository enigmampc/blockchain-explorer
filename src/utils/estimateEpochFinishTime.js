import ethApi from './eth'

const calculateRangesDurations = range => {
  const acc = []

  for (let i = 0; i < range.length; i += 2) {
    const rangeElement = range[i]
    acc.push(rangeElement - range[i + 1])
  }

  return acc
}

const distanceMedian = range => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (length === 0) return length

  durations.sort((a, b) => a - b)

  const half = Math.floor(length / 2)

  return length % 2 ? durations[half] : Math.floor((durations[half - 1] + durations[half]) / 2)
}

const distanceAverage = range => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (!length) return 0

  const sum = durations.reduce((a, b) => a + b)
  return Math.floor(sum / length)
}

const estimateEpochFinishTimes = async (epochs = [], average = false) => {
  let finishTime = 0 // when the current epoch may end
  let finishBlock = 0 // where the current epoch may end
  const currentEpoch = epochs.slice(0, 1)[0] // currently active epoch

  // builds an array of the form [10, 1, 21, 11, 22, 40]
  // where 10 is the last block of the first epoch in the collection
  // and 1 is the first block of the first epoch in the collection
  // 21 is the last block of the second epoch and 11 is it's first block and so on
  const blocksRange = epochs
    .reduce((acc, epoch, index) => {
      const current = index === 0
      if (!current) {
        acc.push(+epoch.startBlockNumber) // start block number for current epoch
        acc.push(+epochs[index - 1].startBlockNumber - 1) // finish block number for current epoch
      }
      return acc
    }, [])
    .reverse()

  try {
    const [currentBlock, epochsRangesTimestamps] = await Promise.all([
      ethApi.getBlockNumber(),
      ethApi.getBatchBlocksTimestamps(blocksRange),
    ])

    const epochBlockCount = average ? distanceAverage(blocksRange) : distanceMedian(blocksRange)
    const epochDuration = average ? distanceAverage(epochsRangesTimestamps) : distanceMedian(epochsRangesTimestamps)

    finishBlock = +currentEpoch.startBlockNumber + epochBlockCount
    const currentBlocksLeft = finishBlock - +currentBlock
    const currentBlocksTimes = epochDuration / epochBlockCount

    finishTime = Math.floor(currentBlocksLeft * currentBlocksTimes) * 1000 // milliseconds
  } catch (e) {
    console.error('Failed to retrieve blocks tiemstamps', e)
  }

  return { finishTime, finishBlock }
}

export default estimateEpochFinishTimes
