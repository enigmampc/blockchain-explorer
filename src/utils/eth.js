import Web3 from 'web3'
// import EnigmaContract from './Enigma.json'
import getNetworkDetailsBy from './networks'

export const ENG_DECIMALS = 8

class EthAPI {
  constructor() {
    if (process.env.REACT_APP_ETH_NETWORK_ID) {
      const { url } = getNetworkDetailsBy('id')(process.env.REACT_APP_ETH_NETWORK_ID)
      this.web3 = new Web3(new Web3.providers.HttpProvider(url))
    } else {
      this.web3 = new Web3(new Web3.providers.HttpProvider(process.env.REACT_APP_ETH_URL))
    }
  }

  getBlockNumber() {
    return this.web3.eth.getBlockNumber()
  }

  async getBlockTimestamp(blockNumber) {
    const block = await this.web3.eth.getBlock(blockNumber)
    return (block && block.timestamp) || '0'
  }

  async getBatchBlocksTimestamps(range) {
    const batch = new this.web3.BatchRequest()

    const blocks = range.map(blockNumber => {
      return new Promise((resolve, reject) => {
        const request = this.web3.eth.getBlock.request(blockNumber, (error, data) => {
          if (error) {
            reject(error)
          } else {
            resolve(data)
          }
        })
        batch.add(request)
      })
    })

    batch.execute()

    return (await Promise.all(blocks)).map(({ timestamp }) => timestamp)
  }
}

export default new EthAPI()
