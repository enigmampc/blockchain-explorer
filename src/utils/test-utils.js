import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import theme from '../theme'

const getStyledComponents = StyledComponent => {
  const componentClass = StyledComponent().type.styledComponentId
  return document.getElementsByClassName(componentClass)
}

const history = createBrowserHistory()

const AllTheProviders = ({ children }) => (
  <ThemeProvider theme={theme}>
    <Router history={history}>{children}</Router>
  </ThemeProvider>
)

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'

export { customRender as render, getStyledComponents, history }
