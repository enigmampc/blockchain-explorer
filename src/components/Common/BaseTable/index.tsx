import React from 'react'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableRow from '@material-ui/core/TableRow'
import { TablePaginationProps } from '@material-ui/core/TablePagination'
import styled from 'styled-components'
import EnhancedTableHead, { EnhancedTableHeadProps } from '../EnhancedTableHead'
import Paginator from '../Paginator'
import Card from '../Card'
import TableOverflow from '../TableOverflow'
import ChevronOpen from './img/chevron-open.svg'
import ChevronClosed from './img/chevron-closed.svg'

type CellProps = {
  id: string
  align: 'left' | 'right' | 'center'
  value: React.ReactNode | string
  status?: 'success' | 'submitted' | 'error'
  colSpan?: number
  style?: object
  onClick?: CallableFunction
}

export type RowProps = {
  id: string
  style?: object
  onClick?: CallableFunction
  cells: CellProps[]
}

export interface BaseTableProps extends React.HTMLAttributes<HTMLDivElement> {
  headerProps?: EnhancedTableHeadProps
  rows: RowProps[]
  paginatorProps?: TablePaginationProps
}

const CardStyled = styled(Card)`
  padding: 0 0 5px 0;
`

const TableCellContent = styled.div`
  display: flex;
`

const TableCellText = styled.span<{ status?: 'success' | 'submitted' | 'error' }>`
  color: ${props => props.theme.status[props.status || 'submitted']};
  font-size: 13px;
  font-weight: normal;
  line-height: 1.2;
  white-space: nowrap;
`

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const Chevron = styled.div<{ isOpen: boolean }>`
  background-image: url(${props => (props.isOpen ? ChevronOpen : ChevronClosed)});
  background-position: 0 50%;
  background-repeat: no-repeat;
  height: 12px;
  margin: 0 12px 0 0;
  transition: all 0.15s linear;
  width: 12px;
`

const BaseTable = ({ headerProps, rows, paginatorProps }: BaseTableProps) => (
  <CardStyled>
    <TableOverflow>
      <Table>
        {headerProps && <EnhancedTableHead {...headerProps} />}
        <TableBody>
          {rows.map(row => (
            <TableRow key={row.id} style={row.style} onClick={() => row.onClick && row.onClick()}>
              {row.cells.map(cell => (
                <TableCell
                  style={{ paddingRight: '16px', ...cell.style }}
                  align={cell.align}
                  key={cell.id}
                  colSpan={cell.colSpan}
                  onClick={() => cell.onClick && cell.onClick()}
                >
                  <TableCellContent>
                    <TableCellText status={cell.status}>{cell.value}</TableCellText>
                  </TableCellContent>
                </TableCell>
              ))}
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableOverflow>
    {paginatorProps && <Paginator {...paginatorProps} />}
  </CardStyled>
)

export default BaseTable
