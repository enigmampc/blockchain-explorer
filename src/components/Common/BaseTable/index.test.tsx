import React from 'react'
import { render, fireEvent } from '../../../utils/test-utils'
import BaseTable, { RowProps } from './'
import { HeaderCell } from '../EnhancedTableHead'

const HEADER_CELLS: HeaderCell[] = [
  { id: '_id', useClassShowOnDesktop: false, align: 'flex-start', label: 'Task ID' },
  { id: 'scAddr', useClassShowOnDesktop: true, align: 'flex-start', label: 'Secret Contact Address' },
  { id: 'workerAddress', useClassShowOnDesktop: true, align: 'flex-start', label: 'Worker Address (Ethereum)' },
]

const ROWS: RowProps[] = [
  {
    id: 'firstRow',
    cells: [
      { id: 'firstCell-firstRow', align: 'center', status: 'success', value: 'first value at first row' },
      { id: 'secondCell-firstRow', align: 'left', status: 'error', value: 'second value at first row' },
      { id: 'thirdCell-firstRow', align: 'right', status: 'submitted', value: 'third value at first row' },
    ],
  },
  {
    id: 'secondRow',
    cells: [
      { id: 'firstCell-secondRow', align: 'center', status: 'success', value: 'first value at second row' },
      { id: 'secondCell-secondRow', align: 'left', status: 'error', value: 'second value at second row' },
      { id: 'thirdCell-secondRow', align: 'right', status: 'submitted', value: 'third value at second row' },
    ],
  },
  {
    id: 'thirdRow',
    cells: [
      { id: 'firstCell-thirdRow', align: 'center', status: 'success', value: 'first value at third row' },
      {
        id: 'thirdCell-thirdRow',
        align: 'right',
        status: 'submitted',
        value: 'third value at third row',
        colSpan: 2,
      },
    ],
  },
]

it(`should render BaseTable without crashing`, () => {
  // Given
  const rows = ROWS.map((row, rI) => {
    row.cells.map((cell, cI) => {
      if (rI === 2 && cI === 1) {
        cell.status = undefined // testing default status property
      }
      return cell
    })

    return row
  })

  // When
  const { container } = render(<BaseTable rows={rows} />)

  // Then
  expect(container).toMatchSnapshot()
})

it(`should render BaseTable with header and footer`, () => {
  // Given
  const headerCells = HEADER_CELLS
  const rows = ROWS

  // When
  const { container } = render(
    <BaseTable
      headerProps={{
        headerCells,
        order: 'asc',
        orderBy: 'id',
        onRequestSort: jest.fn(),
      }}
      rows={rows}
      paginatorProps={{
        colSpan: headerCells.length,
        count: rows.length,
        page: 0,
        rowsPerPage: 1,
        onChangePage: jest.fn(),
      }}
    />,
  )

  // Then
  expect(container).toMatchSnapshot()
})

it(`should render call callbacks in cells`, () => {
  // Given
  const cellCallback = jest.fn()
  const rows = ROWS.map((row, rI) => {
    row.cells.map((cell, cI) => {
      if (rI === 0 && cI === 0) {
        cell.onClick = cellCallback
      }
      return cell
    })

    return row
  })

  // When
  const { getByText } = render(<BaseTable rows={rows} />)
  fireEvent.click(getByText('first value at first row'))

  // Then
  expect(cellCallback).toHaveBeenCalled()
})

it(`should render call callbacks in rows`, () => {
  // Given
  const rowCallback = jest.fn()
  const rows = ROWS.map((row, rI) => {
    if (rI === 0) {
      row.onClick = rowCallback
    }

    return row
  })

  // When
  const { getByText } = render(<BaseTable rows={rows} />)
  fireEvent.click(getByText('first value at first row'))

  // Then
  expect(rowCallback).toHaveBeenCalled()
})
