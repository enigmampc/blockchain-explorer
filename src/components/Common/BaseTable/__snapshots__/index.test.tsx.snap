// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BaseTable with header and footer 1`] = `
.c4 {
  color: #999;
  font-size: 12px;
  white-space: nowrap;
}

.c4 + svg {
  opacity: 1;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
}

.c10 {
  padding: 10px 0 0 0;
}

.c10 > div {
  color: #666;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 12px;
  height: auto;
  min-height: 0;
  padding: 0;
}

.c0 {
  background-color: #fff;
  border-radius: 5px;
  border: solid 1px #efefef;
  box-shadow: 0 0 8px 0 rgba(0,0,0,0.04);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 18px 20px;
}

.c2 {
  color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c3 {
  overflow-x: auto;
  width: 100%;
}

.c1 {
  padding: 0 0 5px 0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c6 {
  color: #4c904a;
  font-size: 13px;
  font-weight: normal;
  line-height: 1.2;
  white-space: nowrap;
}

.c7 {
  color: #954949;
  font-size: 13px;
  font-weight: normal;
  line-height: 1.2;
  white-space: nowrap;
}

.c8 {
  color: #1576ad;
  font-size: 13px;
  font-weight: normal;
  line-height: 1.2;
  white-space: nowrap;
}

@media (min-width:768px) {
  .c9 {
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
  }
}

@media (min-width:768px) {
  .c10 > div {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <table
          class="MuiTable-root"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
                style="padding-right: 5px; display: table-cell;"
              >
                <span
                  class="c4"
                  style="justify-content: flex-start; display: flex; cursor: pointer;"
                >
                  Task ID
                </span>
              </th>
              <th
                class="MuiTableCell-root showOnDesktop MuiTableCell-head"
                scope="col"
                style="padding-right: 5px; display: none;"
              >
                <span
                  class="c4"
                  style="justify-content: flex-start; display: flex; cursor: pointer;"
                >
                  Secret Contact Address
                </span>
              </th>
              <th
                class="MuiTableCell-root showOnDesktop MuiTableCell-head"
                scope="col"
                style="padding-right: 5px; display: none;"
              >
                <span
                  class="c4"
                  style="justify-content: flex-start; display: flex; cursor: pointer;"
                >
                  Worker Address (Ethereum)
                </span>
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c5"
                >
                  <span
                    class="c6"
                  >
                    first value at first row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignLeft MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c5"
                >
                  <span
                    class="c7"
                  >
                    second value at first row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c5"
                >
                  <span
                    class="c8"
                  >
                    third value at first row
                  </span>
                </div>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c5"
                >
                  <span
                    class="c6"
                  >
                    first value at second row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignLeft MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c5"
                >
                  <span
                    class="c7"
                  >
                    second value at second row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c5"
                >
                  <span
                    class="c8"
                  >
                    third value at second row
                  </span>
                </div>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c5"
                >
                  <span
                    class="c6"
                  >
                    first value at third row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-body"
                colspan="2"
                style="padding-right: 16px;"
              >
                <div
                  class="c5"
                >
                  <span
                    class="c8"
                  >
                    third value at third row
                  </span>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="c9"
      >
        <div
          class="MuiTablePagination-root c10"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
          >
            <div
              class="MuiTablePagination-spacer"
            />
            <p
              class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
            >
              Rows per page:
            </p>
            <div
              class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
            >
              <select
                aria-label="rows per page"
                class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiInputBase-input MuiInputBase-inputSelect"
              >
                <option
                  class="MuiTablePagination-menuItem"
                  value="25"
                >
                  25
                </option>
                <option
                  class="MuiTablePagination-menuItem"
                  value="50"
                >
                  50
                </option>
                <option
                  class="MuiTablePagination-menuItem"
                  value="100"
                >
                  100
                </option>
                <option
                  class="MuiTablePagination-menuItem"
                  value="200"
                >
                  200
                </option>
              </select>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <p
              class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
            >
              1-1 of 3
            </p>
            <div
              style="flex-shrink: 0;"
            >
              <button
                aria-label="first page"
                class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                    />
                    <path
                      d="M24 24H0V0h24v24z"
                      fill="none"
                    />
                  </svg>
                </span>
              </button>
              <button
                aria-label="previous page"
                class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
                    />
                  </svg>
                </span>
              </button>
              <button
                aria-label="next page"
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="last page"
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render BaseTable without crashing 1`] = `
.c0 {
  background-color: #fff;
  border-radius: 5px;
  border: solid 1px #efefef;
  box-shadow: 0 0 8px 0 rgba(0,0,0,0.04);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 18px 20px;
}

.c2 {
  color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c3 {
  overflow-x: auto;
  width: 100%;
}

.c1 {
  padding: 0 0 5px 0;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  color: #4c904a;
  font-size: 13px;
  font-weight: normal;
  line-height: 1.2;
  white-space: nowrap;
}

.c6 {
  color: #954949;
  font-size: 13px;
  font-weight: normal;
  line-height: 1.2;
  white-space: nowrap;
}

.c7 {
  color: #1576ad;
  font-size: 13px;
  font-weight: normal;
  line-height: 1.2;
  white-space: nowrap;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <table
          class="MuiTable-root"
        >
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c4"
                >
                  <span
                    class="c5"
                  >
                    first value at first row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignLeft MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c4"
                >
                  <span
                    class="c6"
                  >
                    second value at first row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c4"
                >
                  <span
                    class="c7"
                  >
                    third value at first row
                  </span>
                </div>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c4"
                >
                  <span
                    class="c5"
                  >
                    first value at second row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignLeft MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c4"
                >
                  <span
                    class="c6"
                  >
                    second value at second row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c4"
                >
                  <span
                    class="c7"
                  >
                    third value at second row
                  </span>
                </div>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-body"
                style="padding-right: 16px;"
              >
                <div
                  class="c4"
                >
                  <span
                    class="c5"
                  >
                    first value at third row
                  </span>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-body"
                colspan="2"
                style="padding-right: 16px;"
              >
                <div
                  class="c4"
                >
                  <span
                    class="c7"
                  >
                    third value at third row
                  </span>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
